Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp

```diff
@@ -558,6 +571,153 @@ void HTMLScriptElement::inserted()
+    // 6. If parsed["scopes"] exists, then:
+    constexpr StringView scopes_key = "scopes"sv;
+    if (parsed_obj.has(scopes_key)) {
+        // 1. If parsed["scopes"] is not an ordered map, then throw a TypeError indicating that the value for the "scopes" top-level key needs to be a JSON object.
+        if (!parsed_obj.has_object(scopes_key)) {
+            dbgln("TypeError: the value for \"scopes\" top-level key needs to be a JSON object");
+            return sorted_and_normalized;
+        }
+        // 2. Set sortedAndNormalizedScopes to the result of sorting and normalizing scopes given parsed["scopes"] and baseURL.
+        sorted_and_normalized_scopes = TRY(
+            sort_and_normalize_scopes(parsed_obj.get_object(scopes_key).release_value()));
+    }
+
+    return sorted_and_normalized;
+}
+
+// https://html.spec.whatwg.org/multipage/webappapis.html#sorting-and-normalizing-a-module-specifier-map
+ErrorOr<ModuleSpecifierMap> HTMLScriptElement::sort_and_normalize_module_specifier_map(JsonObject const& import_obj) const
+{
+    // 1. Let normalized be an empty ordered map
+    ModuleSpecifierMap normalized;
+    // 2. For each specifierKey → value of originalMap
+    import_obj.for_each_member([this, &normalized](DeprecatedString const& specifier_key, JsonValue const& value) {
+        // 1. Let normalizedSpecifierKey be the result of normalizing a specifier key given specifierKey and baseURL.
+        auto const normalized_specifier_key = normalize_a_specifier_key(specifier_key);
+        // 2. If normalizedSpecifierKey is null, then continue.
+        if (!normalized_specifier_key.has_value()) {
+            return;
+        }
+        // 3. If value is not a string, then:
+        if (!value.is_string()) {
+            // 1. FIXME The user agent may report a warning to the console indicating that addresses need to be strings.
+            dbgln("addresses need to be strings");
+            // 2. Set normalized[normalizedSpecifierKey] to null.
+            normalized.set(normalized_specifier_key.value(), OptionalNone());
+            // 3. Continue
+            return;
+        }
+        // 4. Let addressURL be the result of resolving a URL-like module specifier given value and baseURL.
+        AK::URL const address_url = resolve_a_url_like_module_specifier(value.as_string());
+        // 5. If addressURL is null, then:
+        if (!address_url.is_valid()) {
+            // 1. FIXME The user agent may report a warning to the console indicating that the address was invalid.
+            dbgln("addressURL was invalid");
+            // 2. Set normalized[normalizedSpecifierKey] to null
+            normalized.set(normalized_specifier_key.value(), OptionalNone());
+            // 3. Continue
+            return;
+        }
+        // 6. If specifierKey ends with U+002F (/), and the serialization of addressURL does not end with U+002F (/), then
+        if (specifier_key.ends_with('/') && address_url.basename().ends_with('/')) {
```

Comment:
Two things:
- Should be checking the serialization of address_url, not its basename.
- The condition is inverted, we want to check it *does not* end with `/`.